diff --git a/pycnn/pycnn.pyx b/pycnn/pycnn.pyx
index ac9aba9..2036c5d 100644
--- a/pycnn/pycnn.pyx
+++ b/pycnn/pycnn.pyx
@@ -35,7 +35,7 @@ cimport pycnn
 cdef init(random_seed=None):
     cdef int argc = len(sys.argv)
     cdef char** c_argv
-    args = [bytes(x) for x in sys.argv]
+    args = [bytes(x,'utf-8') for x in sys.argv]
     c_argv = <char**>malloc(sizeof(char*) * len(args)) # TODO check failure?
     for idx, s in enumerate(args):
         c_argv[idx] = s
@@ -402,9 +402,6 @@ cdef class Expression: #{{{
     def __getitem__(self, int i):
         return pick(self, i)
 
-    def __getslice__(self, int i, int j):
-        return pickrange(self, i, j)
-
     cpdef scalar_value(self, recalculate=False):
         if self.cg_version != _cg._cg_version: raise RuntimeError("Stale Expression (created before renewing the Computation Graph).")
         if recalculate: self.cg().forward()
